Add blog post - Implement custom source connector for Table API and SQL (#460)
apache/flink-web: _posts/2021-08-16-connector-table-sql-api-part1.md

---
layout: post

> title:  "Implementing a custom source connector for Table API and SQL - Part One "

Reviewer: Should we do tile style `Implementing a Custom Source Connector for Table API and SQL`?

date: 2021-08-18T00:00:00.000Z
authors:
- Ingo Buerk:
  name: "Ingo Buerk"
excerpt: 
---

{% toc %}

> # Introduction

Reviewer 2: As a general remark, I know that in the documentation we use "you" rather than "we". This is of course a completely biased opinion, but in a blog post like this "we" reads much more inclusive to me compared to "You will be doing XYZ". I'm wondering how others feel about this?

Author: I based the decision on some previous experiences at other companies and also on https://www.gatsbyjs.com/contributing/gatsby-style-guide/#word-choice but I am open to a discussion about it!

Reviewer 2: It does seem to be a common way of doing it, so I guess "it's just me". :-)

> Apache Flink is a data processing engine that keeps [state](https://ci.apache.org/projects/flink/flink-docs-release-1.13/docs/ops/state/state_backends/) locally in order to do computations but does not store data. This means that it does not include its own fault-tolerant storage component by default and relies on external systems to ingest and persist data. Connecting to external data input (**sources**) and external data storage (**sinks**) is achieved with interfaces called **connectors**.   

Reviewer: ```suggestion
Apache Flink is a data processing engine that aims to keep [state](https://ci.apache.org/projects/flink/flink-docs-release-1.13/docs/ops/state/state_backends/) locally in order to do computations efficiently. However, Flink does not "own" the data but relies on external systems to ingest and persist data. Connecting to external data input (**sources**) and external data storage (**sinks**) is usually summarized under the term **connectors** in Flink.   
```

Reviewer: I found the previous wording very negative.

> Since connectors are such important components, Flink ships with [connectors for some popular systems](https://ci.apache.org/projects/flink/flink-docs-release-1.13/docs/connectors/table/overview/). But sometimes you may need to read in an uncommon data format and what Flink provides is not enough. This is why Flink also provides [APIs](#) for building custom connectors if you want to connect to a system that is not supported by an existing connector.   

Reviewer: ```suggestion
Since connectors are such important components, Flink ships with [connectors for some popular systems](https://ci.apache.org/projects/flink/flink-docs-release-1.13/docs/connectors/table/overview/). But sometimes you may need to read in an uncommon data format and what Flink provides is not enough. This is why Flink also provides [extension points](??????) for building custom connectors if you want to connect to a system that is not supported by an existing connector.   
```

> Once you have a source and a sink defined for Flink, you can use its declarative APIs (in the form of the [Table API and SQL](https://ci.apache.org/projects/flink/flink-docs-release-1.13/docs/dev/table/overview/)) to execute queries for data analysis without modification to the underlying data.  

Reviewer: > without modification to the underlying data

for sinks your are modifying data, maybe drop `and a sink`

> The **Table API** has the same operations as **SQL** but it extends and improves SQL's functionality. It is named Table API because of its relational functions on tables: how to obtain a table, how to output a table, and how to perform query operation on the table.

Reviewer 2: The first sentence here sounds a bit like SQL is inferior to Table API. I think we should avoid that and treat them as equal. Table API provides a more programmatic access while SQL is, well, SQL, which is a more universal language.

Reviewer: +1

> In this two-part tutorial, you will explore some of these APIs and concepts by implementing your own custom source connector for reading in data from a mailbox. You will use Flink to process an email inbox through the IMAP protocol and sort them out by subject to a sink. 

Reviewer 2: I really like the idea of presenting an achievable goal for the end! But reading this for the first time I'm not sure what "sorting them out by subject to a sink" means.

Reviewer 2: Especially if we do really split this into two blog posts, how about having some "teaser" here of what the final result will look like?

Reviewer 2: Having read the entirety now, we never actually achieved this goal? Did I misunderstand the intention here?

Reviewer: link to the IMAP wikipedia article?

Author: I will fix this.  I was also learning about connectors as I was doing this.  :laughing:

Part one will focus on building a custom source connector and [part two](#) will focus on integrating it. 

# Goals

Part one of this tutorial will teach you how to build and run a custom source connector to be used with Table API and SQL, two high-level abstractions in Flink.

You are encouraged to follow along with the code in this [repository](github.com/Airblader/blog-imap). It provides a boilerplate project that also comes with a bundled [docker-compose](https://docs.docker.com/compose/) setup that lets you easily run the connector. You can then try it out with Flink’s SQL client.

# Prerequisites

This tutorial assumes that you have some familiarity with Java and objected-oriented programming. 

It would also be useful to have [docker-compose](https://docs.docker.com/compose/install/) installed on your system in order to use the script included in the repository that builds and runs the connector. 

# Understand the infrastructure required for a connector

In order to create a connector which works with Flink, you need:

> 1. A _factory class_ (a blueprint for creating other objects) that tells Flink with which identifier (in this case, “imap”) our connector can be addressed, which configuration options it exposes, and how the connector can be instantiated. Since Flink uses the Java Service Provider Interface (SPI) to discover factories located in different modules, you will also need to add some configuration details.

Reviewer: ```suggestion
1. A _factory class_ (a blueprint for creating other objects from string properties) that tells Flink with which identifier (in this case, “imap”) our connector can be addressed, which configuration options it exposes, and how the connector can be instantiated. Since Flink uses the Java Service Provider Interface (SPI) to discover factories located in different modules, you will also need to add some configuration details.
```

> 2. The _table source_ object as a specific instance of the connector during the planning stage. It tells Flink some information about this instance and how it can create the connector runtime implementation. There are also more advanced features, such as [abilities](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/table/connector/source/abilities/package-summary.html), that can be implemented to improve connector performance.

Reviewer: mention that is responsible for  back and forth communicating with the optimizer during planning and is kind of a another factory for creating connector runtime implementation

3. A _runtime implementation_ from the connector obtained during the planning stage. The runtime logic is implemented in Flink's core connector interfaces and does the actual work of producing rows of dynamic table data. The runtime instances are shipped to the Flink cluster. 

> Let us look at  this sequence (factory class → table source → runtime implementation) in reverse order.

Reviewer: nit: double space

# Establish the runtime implementation of the connector

You first need to have a source connector which can be used in Flink's runtime system, defining how data goes in and how it can be executed in the cluster. There are a few different interfaces available for implementing the actual source of the data and have it be discoverable in Flink.  

> For complex connectors, you may want to implement the [Source interface](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/api/connector/source/Source.html) which gives you a lot of control. For simpler use cases, you can use the [SourceFunction interface](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/streaming/api/functions/source/SourceFunction.html), which is the base interface for all stream data sources in Flink. There are already a few different implementations of SourceFunction interfaces for common use cases such as the [FromElementsFunction](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/streaming/api/functions/source/FromElementsFunction.html) class and the [RichSourceFunction](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/streaming/api/functions/source/RichSourceFunction.html) class. You will use the latter.  

Reviewer: for clarity, `Source` interface is the new abstraction whereas `SourceFunction` is the old. all connectors will implement `Source` eventually but currently more simplification is needed to expose it to users.

> which is the base interface for all stream data sources in Flink

this is not true, Kafka and Filesystem are already using the new abstraction

> `RichSourceFunction` is a base class for implementing a parallel data source that has access to context information and some lifecycle methods. There is a `run()` method inherited from the `SourceFunction` interface that you need to implement. It is invoked once and can be used to produce the data either once for a bounded result or within a loop for an unbounded stream.

Reviewer 2: The "parallel" is wrong here, that'd be `RichParallelSourceFunction` (which we're not using)

> `RichSourceFunction` is a base class for implementing a parallel data source that has access to context information and some lifecycle methods. There is a `run()` method inherited from the `SourceFunction` interface that you need to implement. It is invoked once and can be used to produce the data either once for a bounded result or within a loop for an unbounded stream.

Reviewer 2: I'm not sure about the description of bounded / unbounded here. Bounded really just means that the number of produced records is known to be finite, while unbounded sources can produce infinitely many records. This means bounded sources finish, while unbounded ones do not. There are more implications of this, but the point is that bounded sources can certainly run "in a loop" and for long times as well, it's just that they will eventually finish.

For example, to create a bounded data source, you could implement this method so that it reads all existing emails and then closes. To create an unbounded source, you could only look at new emails coming in while the source is active. You can also combine these behaviors and expose them through configuration options.

When you first create the class and implement the interface, it should look something like this:

```java
import org.apache.flink.streaming.api.functions.source.RichSourceFunction;

> import org.apache.flink.table.data.RowData;

Reviewer: mention that we use internal data structures because that is required by the table runtime

This is what a scan table source implementation would look like:

```java
import org.apache.flink.table.connector.ChangelogMode;
import org.apache.flink.table.connector.source.DynamicTableSource;
import org.apache.flink.table.connector.source.ScanTableSource;
import org.apache.flink.table.connector.source.SourceFunctionProvider;

public class ImapTableSource implements ScanTableSource {
  @Override
  public ChangelogMode getChangelogMode() {
    return ChangelogMode.insertOnly();
  }

  @Override
  public ScanRuntimeProvider getScanRuntimeProvider(ScanContext ctx) {

>     boolean bounded = true;

Reviewer 2: Hm, I guess you extracted this `true` to give it a name, but as a developer, this local variable looks useless and a code review would probably point that out. IDEs will typically name the argument for you anyway, i.e. `of(…, true)` will be shown as `of(…, bounded: true)`.

>     boolean bounded = true;

Reviewer 2: (At the very least we should also make it `final boolean bounded`, but I'm really not sure about keeping it as a variable)

Reviewer: for a tutorial a local variable makes the code more readable

Reviewer 2: Fair enough. I meant to delete this thread but apparently that doesn't work, so my original comment is gone now. We can just ignore this thread. :-)

    final ImapSourceFunction sourceFunction = new ImapSourceFunction();
    return SourceFunctionProvider.of(sourceFunction, bounded);
  }

  @Override
  public DynamicTableSource copy() {
    return new ImapTableSource();
  }

  @Override
  public String asSummaryString() {
    return "IMAP Table Source";
  }
}
```

[ChangelogMode](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/table/connector/ChangelogMode.html) informs Flink of expected changes that the planner can expect during runtime. For example, whether the source produces only new rows, also updates to existing ones, or whether it can remove previously produced rows. Our source will only produce (`insertOnly()`) new rows.

> [ScanRuntimeProvider](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/table/connector/source/ScanTableSource.ScanRuntimeProvider.html) allows Flink to create the actual runtime implementation you established previously (for reading the data). Flink even provides utilities like [SourceFunctionProvider](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/table/connector/source/SourceFunctionProvider.html) to wrap it into an instance of [SourceFunction](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/streaming/api/functions/source/SourceFunction.html), which is the base unit of all stream data sources in Flink.  

Reviewer: > which is the base unit of all stream data sources in Flink

not true, better "one of the base runtime interfaces"

    factoryHelper.validate();

    return new ImapTableSource();
  }
}
```

There are currently no configuration options but they can be added and also validated within the `createDynamicTableSource()` function. There is a small helper utility, [TableFactoryHelper](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/table/factories/FactoryUtil.TableFactoryHelper.html), that Flink offers which ensures that required options are set and that no unknown options are provided.

Finally, you need to register your factory for Java's Service Provider Interfaces (SPI). Classes that implement this interface can be discovered and should be added to this file `src/main/resources/META-INF/services/org.apache.flink.table.factories.Factory` with the fully classified class name of your factory:

```java
// if you created your class in the package org.example.acme, it should be named the following:
org.example.acme.ImapTableSourceFactory
```

# Test the custom connector

You should now have a working source connector. If you are following along with the provided repository, you can test it by running:

```sh
$ cd testing/
$ ./build_and_run.sh
```

> This builds the connector, starts a Flink cluster, a [test email server](https://greenmail-mail-test.github.io/greenmail/) which you will need later), and the [SQL client](https://ci.apache.org/projects/flink/flink-docs-release-1.13/docs/dev/table/sqlclient/) for you. You may need to run this script as a super user. If successful, you should see the SQL CLI:

Reviewer 2: There's a missing `(` for the "which you will need later", right?

> This builds the connector, starts a Flink cluster, a [test email server](https://greenmail-mail-test.github.io/greenmail/) which you will need later), and the [SQL client](https://ci.apache.org/projects/flink/flink-docs-release-1.13/docs/dev/table/sqlclient/) for you. You may need to run this script as a super user. If successful, you should see the SQL CLI:

Reviewer 2: > You may need to run this script as a super user.

I would refrain from this piece of advice in a blog post. Running docker containers as root is a bad idea. Yes, people can get stuck because they have no proper docker setup on their machine, and we don't have the time in a blog post to explain all of that, but teaching people to give random scripts or docker containers root access is wrong.

I could help here on the script side by catching this case and printing a helpful error message instead.

> ![image](https://ci.apache.org/projects/flink/flink-docs-release-1.13/fig/sql_client_demo.gif)

Reviewer 2: I'm not sure how relevant this is, but this GIF contains `SHOW TABLES` showing `TaxiRides`, which wouldn't happen in the actual setup.

Author: Haha you're right... I was trying to do **DRY** but wasn't careful here. :see_no_evil:

> You can now create a table with your connector by executing the following statement with the SQL client:

Reviewer: was there any link to docs and explanation what is SQL Client before?

```sql
CREATE TABLE T (subject STRING, content STRING) WITH ('connector' = 'imap');

SELECT * FROM T;
```

*Note:* The schema must be exactly as written since it is currently hardcoded into the connector.

> You should be able to see some static data.  

Reviewer 2: nit: "You should be able to see the static data we provided in our source connector earlier."?

---
layout: post
title:  "Implementing a Custom Source Connector for Table API and SQL - Part One "

> date: 2021-08-27T00:00:00.000Z

Reviewer 2: Do we need to adjust the publishing date again? Also, the file names of the two blog posts are out of sync with this.

> authors:

Reviewer 2: I'm new to working with a technical writer, but with the amount of work you have contributed to this at this point as well, does it make sense to add your name as well @infoverload?

Author: Oh that would be nice!

- Ingo Buerk:
  name: "Ingo Buerk"
excerpt: 
---

{% toc %} 

# Introduction

Apache Flink is a data processing engine that aims to keep [state](https://ci.apache.org/projects/flink/flink-docs-release-1.13/docs/ops/state/state_backends/) locally in order to do computations efficiently. However, Flink does not "own" the data but relies on external systems to ingest and persist data. Connecting to external data input (**sources**) and external data storage (**sinks**) is usually summarized under the term **connectors** in Flink. 

> Since connectors are such important components, Flink ships with [connectors for some popular systems](https://ci.apache.org/projects/flink/flink-docs-release-1.13/docs/connectors/table/overview/). But sometimes you may need to read in an uncommon data format and what Flink provides is not enough. This is why Flink also provides [extension points](#) for building custom connectors if you want to connect to a system that is not supported by an existing connector.   

Reviewer 2: "extension points" doesn't link to anything, and I'm not sure what it was intended to link to?

Author: Is there a good page I should link it to?

Once you have a source and a sink defined for Flink, you can use its declarative APIs (in the form of the [Table API and SQL](https://ci.apache.org/projects/flink/flink-docs-release-1.13/docs/dev/table/overview/)) to execute queries for data analysis.  

The **Table API** provides more programmatic access while **SQL** is a more universal query language. It is named Table API because of its relational functions on tables: how to obtain a table, how to output a table, and how to perform query operations on the table.

In this two-part tutorial, you will explore some of these APIs and concepts by implementing your own custom source connector for reading in data from an email inbox. You will then use Flink to process emails through the [IMAP protocol](https://en.wikipedia.org/wiki/Internet_Message_Access_Protocol). 

Part one will focus on building a custom source connector and [part two](#) will focus on integrating it. 

# Goals

Part one of this tutorial will teach you how to build and run a custom source connector to be used with Table API and SQL, two high-level abstractions in Flink.

> You are encouraged to follow along with the code in this [repository](github.com/TNG/flink-connector-email). It provides a boilerplate project that also comes with a bundled [docker-compose](https://docs.docker.com/compose/) setup that lets you easily run the connector. You can then try it out with Flink’s SQL client.

Reviewer 2: There are actually two repositories, one "shell" repository to follow along with, and the more complete implementation:

* https://github.com/Airblader/blog-imap for following along
* https://github.com/TNG/flink-connector-email for an implementation

Here we want to link to the former, not the latter, plus the missing `https://` breaks the link. Most of the suggestions following this one are the same two issues, so I'll only explain it here once.

```suggestion
You are encouraged to follow along with the code in this [repository](https://github.com/Airblader/blog-imap). It provides a boilerplate project that also comes with a bundled [docker-compose](https://docs.docker.com/compose/) setup that lets you easily run the connector. You can then try it out with Flink’s SQL client.
```

When you first create the class and implement the interface, it should look something like this:

```java
import org.apache.flink.streaming.api.functions.source.RichSourceFunction;
import org.apache.flink.table.data.RowData;

public class ImapSource extends RichSourceFunction<RowData> {
  @Override
  public void run(SourceContext<RowData> ctx) throws Exception {}

  @Override
  public void cancel() {}
}
```

Note that internal data structures (`RowData`) are used because that is required by the table runtime.

In the `run()` method, you get access to a [context](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/streaming/api/functions/source/SourceFunction.SourceContext.html) object inherited from the SourceFunction interface, which is a bridge to Flink and allows you to output data. Since the source does not produce any data yet, the next step is to make it produce some static data in order to test that the data flows correctly: 

```java
import org.apache.flink.streaming.api.functions.source.RichSourceFunction;
import org.apache.flink.table.data.GenericRowData;
import org.apache.flink.table.data.RowData;

> import org.apache.flink.table.data.StringData

Reviewer 2: ```suggestion
import org.apache.flink.table.data.StringData;
```

This is what a scan table source implementation would look like:

```java
import org.apache.flink.table.connector.ChangelogMode;
import org.apache.flink.table.connector.source.DynamicTableSource;
import org.apache.flink.table.connector.source.ScanTableSource;
import org.apache.flink.table.connector.source.SourceFunctionProvider;

public class ImapTableSource implements ScanTableSource {
  @Override
  public ChangelogMode getChangelogMode() {
    return ChangelogMode.insertOnly();
  }

  @Override
  public ScanRuntimeProvider getScanRuntimeProvider(ScanContext ctx) {
    boolean bounded = true;

>     final ImapSourceFunction sourceFunction = new ImapSourceFunction();
>     return SourceFunctionProvider.of(sourceFunction, bounded);

Reviewer 2: ```suggestion
    final ImapSource source = new ImapSource();
    return SourceFunctionProvider.of(source, bounded);
```

[ScanRuntimeProvider](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/table/connector/source/ScanTableSource.ScanRuntimeProvider.html) allows Flink to create the actual runtime implementation you established previously (for reading the data). Flink even provides utilities like [SourceFunctionProvider](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/table/connector/source/SourceFunctionProvider.html) to wrap it into an instance of [SourceFunction](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/streaming/api/functions/source/SourceFunction.html), which is one of the base runtime interfaces.

You will also need to indicate whether the source is bounded or not. Currently, this is the case but you will have to change this later. 

# Create a factory class for the connector so it can be discovered by Flink

You now have a working source connector, but in order to use it in Table API or SQL, it needs to be discoverable by Flink. You also need to define how the connector is addressable from a SQL statement when creating a source table. 

> You need to implement a [Factory](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/table/factories/Factory.html), which is a base interface that create object instances from a list of key-value pairs in Flink's Table API and SQL.  A factory is uniquely identified by its class name and `factoryIdentifier()`.  For this tutorial, you will implement the more specific [DynamicTableSourceFactory](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/table/factories/DynamicTableSourceFactory.html), which allows you to configure a dynamic table connector as well as create `DynamicTableSource` instances.  

Reviewer 2: ```suggestion
You need to implement a [Factory](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/table/factories/Factory.html), which is a base interface that creates object instances from a list of key-value pairs in Flink's Table API and SQL.  A factory is uniquely identified by its class name and `factoryIdentifier()`.  For this tutorial, you will implement the more specific [DynamicTableSourceFactory](https://ci.apache.org/projects/flink/flink-docs-master/api/java/org/apache/flink/table/factories/DynamicTableSourceFactory.html), which allows you to configure a dynamic table connector as well as create `DynamicTableSource` instances.  
```

```java

> import java.util.HashSet;

Reviewer 2: ```suggestion
import java.util.HashSet;
import java.util.Set;
```

	<p class="align-center">Flink SQL Client</p>
</div>

You can now create a table (with a "subject" column and a "content" column) with your connector by executing the following statement with the SQL client:

```sql
CREATE TABLE T (subject STRING, content STRING) WITH ('connector' = 'imap');

SELECT * FROM T;
``` 

*Note:* The schema must be exactly as written since it is currently hardcoded into the connector.

You should be able to see the static data you provided in your source connector earlier, which would be "Subject 1" and "Hello, World!".

Now that you have a working connector, the next step is to make it do something more useful than returning static data.

# Summary

In this tutorial, you looked into the infrastructure required for a connector and configured its runtime implementation to define how it should be executed in a cluster. You also defined a dynamic table source that reads the entire stream-converted table from the external source, made the connector discoverable by Flink through creating a factory class for it, and then tested it.  

# Next Steps

> In [part two](#), you will integrate this connector with an email inbox through the IMAP protocol. 

Reviewer 2: ```suggestion
In [part two](/2021/08/27/connector-table-sql-api-part2), you will integrate this connector with an email inbox through the IMAP protocol. 
```

(However, the link might change again if we change the dates)
